Fix cast always succeeding due to Godot bug.

joshua-maros opened the pull request:
See #158 . 

Currently, casting always succeeds due to a bug in Godot. This PR adds an explicit (slow) check to make sure that the cast is valid before performing it. This should mirror the behavior that Godot will have when the bug is fixed upstream. This is better than the fix mentioned in #158 as that fix only works when the object you are casting is *exactly* the class you are casting to, you could not cast a derived class to a parent class. This PR has no such limitation.

I tested it with the following code:

```rust
#[derive(GodotClass)]
#[class(init)]
pub struct TestClass {}

#[godot_api]
impl TestClass {}

#[derive(GodotClass)]
#[class(init)]
pub struct OtherRefCountedClass {}

#[godot_api]
impl OtherRefCountedClass {}

#[derive(GodotClass)]
#[class(init)]
pub struct TestClassRunner {
    #[export]
    #[init(default = Gd::new(TestClass {}))]
    pub tc: Gd<TestClass>,
}

#[godot_api]
impl TestClassRunner {
    #[func]
    pub fn tests(&self) {
        godot_print!("{:?}", self.tc.share().upcast::<Object>());
        godot_print!("{:?}", self.tc.share().upcast::<RefCounted>());
        godot_print!(
            "{:?}",
            self.tc.share().upcast::<Object>().cast::<TestClass>()
        );
        godot_print!(
            "{:?}",
            self.tc.share().upcast::<Object>().try_cast::<Node>()
        );
        godot_print!(
            "{:?}",
            self.tc
                .share()
                .upcast::<Object>()
                .try_cast::<OtherRefCountedClass>()
        );
    }
}
```

And constructing the runner and running the function in Godot produced the following output:
```
Gd { id: -9223372008501280223, class: TestClass }
Gd { id: -9223372008501280223, class: TestClass }
Gd { id: -9223372008501280223, class: TestClass }
None
None
```

joshua-maros commented:
I've incorporated your feedback now.

sayaks commented on itest/rust/src/object_test.rs:
```diff
@@ -416,6 +417,18 @@ fn object_engine_downcast() {
+    node3d.set_position(pos);
+
+    let object = node3d.share().upcast::<Object>();
+    assert!(object.try_cast::<Node2D>().is_none());
```
Doesn't this already fail without this bug fix? From what i remember, #158 only happens when using custom classes, so you'd need to make one/two custom class(es) here to test this downcasting issue properly right?

sayaks reviewed (commented):
thank you!

looking through these tests, i dont think we have any tests that downcasting succeeds for custom classes when it should either. That wasn't an issue before but it would be good to have for regression. Could you add that too?

joshua-maros commented:
I've changed the tests accordingly.

I also found out the variant conversion logic doesn't go through the casting logic so it has the same problem. I've added a fix w/ todo comments and a test. It requires an extra trait bound but my understanding is that everything that implements `GodotClass` also implements `Inherits<Object>` so it shouldn't cause problems.

sayaks commented:
oo great catch! and yeah the `Inherits<Object>` shouldn't be an issue usually.

bors try

Bromeon commented:
Thanks a lot! 👍 

Could you name the `TODO` as `TODO(#234)` -- even if this is not the tracking issue, it's good to relate them somehow, so we can find them later and remove them all at once.

Out of curiosity, I guess the `try_cast` requires the `Inherits<Object>` bound? AFAIR the inner cast methods don't have that bound, would it be possible to use those or would it lead to code duplication?

Bromeon commented:
In addition to my last remarks, could you address the small merge conflict? Sorry for the inconvenience.

joshua-maros commented:
- Updated TODOs and resolved conflict

You're right, the internal method does not require the bound, and I see no problem with making `try_from_variant` use that instead.

Bromeon reviewed (approved):
Thanks a lot!

bors r+
